libbeat/processors/annotate/kubernetes/indexing.go

```diff
@@ -0,0 +1,262 @@
+package kubernetes
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+
+	corev1 "github.com/ericchiang/k8s/api/v1"
+)
+
+// Indexing is the singleton Register instance where all Indexers and Matchers
+// are stored
+var Indexing = NewRegister()
+
+// Indexer take known pods and generate all the metadata we need to enrich
+// events in a efficient way. By preindexing the metadata in the way it will be
+// checked when matching events
+type Indexer interface {
+	// GetMetadata generates event metadata for the given pod, then returns the
+	// list of indexes to create, with the metadata to put on them
+	GetMetadata(pod *corev1.Pod) []MetadataIndex
+
+	// GetIndexes return the list of indexes the given pod belongs to. This function
+	// must return the same indexes than GetMetadata
+	GetIndexes(pod *corev1.Pod) []string
+}
+
+// MetadataIndex holds a pair of index -> metadata info
+type MetadataIndex struct {
+	Index string
+	Data  common.MapStr
+}
+
+// Matcher takes a new event and returns the index
+type Matcher interface {
+	// MetadataIndex returns the index string to use in annotation lookups for the given
+	// event. A previous indexer should have generated that index for this to work
+	// This function can return "" if the event doesn't match
+	MetadataIndex(event common.MapStr) string
+}
+
+type Indexers struct {
+	sync.RWMutex
+	indexers []Indexer
+}
+
+type Matchers struct {
+	sync.RWMutex
+	matchers []Matcher
+}
+
+// Register contains Indexer and Matchers to use on pod indexing and event matching
+type Register struct {
+	indexers map[string]IndexConstructor
+	matchers map[string]MatcherConstructor
+
+	defaultIndexers []Indexer
+	defaultMatchers []Matcher
+}
+
+type IndexConstructor func(config common.Config) (Indexer, error)
+type MatcherConstructor func(config common.Config) (Matcher, error)
+
+// NewRegister creates and returns a new Register.
+func NewRegister() *Register {
+	return &Register{
+		indexers: make(map[string]IndexConstructor, 0),
+		matchers: make(map[string]MatcherConstructor, 0),
+
+		defaultIndexers: make([]Indexer, 0),
+		defaultMatchers: make([]Matcher, 0),
+	}
+}
+
+// AddIndexer to the register
+func (r *Register) AddIndexer(name string, indexer IndexConstructor) {
+	r.indexers[name] = indexer
```

Comment:
I'm missing all RWMutex locking logic in getters & setters, could you please add it? Like in https://github.com/elastic/beats/pull/3888/files#diff-2afe85a18185c27752efda431319dac3R120